Remove IOperation laziness from SemanticModel

333fred opened the pull request:
Previously, a `ConcurrentDictionary`s existed on the operation factories, maintaining a `BoundNode`->`Operation` mapping that was responsible for ensuring that all returned nodes are the same. I've removed this cache, and rewritten how the semantic models fetch operations. Now, they get the bound node for the root of the semantic model, and get the operation it. They then build a cache from `SyntaxNode`->explicit `Operation`, which can be quickly queried by future calls to `GetOperation`. As part of this, I added a new public `OperationWalker<TArgument>` type, which walks an operation tree given an argument, and is used for sharing instances of the cache populater. Commit 1 enables C#, commit 2 enables VB.

333fred commented on src/Compilers/CSharp/Portable/Compilation/InitializerSemanticModel.cs:
```diff
@@ -90,24 +90,23 @@ internal override BoundNode GetBoundRoot()
             {
                 case SyntaxKind.VariableDeclarator:
-                    rootSyntax = ((VariableDeclaratorSyntax)rootSyntax).Initializer.Value;
+                    rootSyntax = ((VariableDeclaratorSyntax)rootSyntax).Initializer;
```
`GetBoundRoot` is only used by the region analysis APIs. This change ensures that we have a bound node for the equals value, which ensures that the SemanticModel will generate the correct the correct `ISymbolInitializerOperation` nodes for the current context, and the extra node does not impact the results of the region analysis apis.

333fred commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -351,6 +316,20 @@ internal IOperation CreateInternal(BoundNode boundNode)
             }
         }
 
+        public ImmutableArray<TOperation> CreateFromArray<TBoundNode, TOperation>(ImmutableArray<TBoundNode> boundNodes) where TBoundNode : BoundNode where TOperation : class, IOperation
```
This is just a move from above. No implementation changes.

333fred commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1946,13 +1925,21 @@ private ILabeledOperation CreateBoundLabeledStatementOperation(BoundLabeledState
+
+            // If we're creating the tree for a speculatively-bound constructor initializer, there can be a bound sequence as the child node here
+            // that corresponds to the lifetime of any declared variables.
+            IOperation expression = Create(boundExpressionStatement.Expression);
```
This is tested by the added verification to `Scope_ConstructorInitializers_06`.

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6982,12 +6982,7 @@ private void VisitUsingVariableDeclarationOperation(IUsingDeclarationOperation o
-                ((Operation)operation).OwningSemanticModel,
-                operation.Syntax,
-                isImplicit: true);
+            BlockOperation logicalBlock = BlockOperation.CreateTemporaryBlock(statements, ((Operation)operation).OwningSemanticModel!, operation.Syntax);
```
This is required because of the uncommented verification code in Operation.SetParentOperation. It was always failing, but we had commented it out because of race conditions, and so missed that this was an issue when it was added.

AlekseyTs replied:
> This is required because of the uncommented verification code in Operation.SetParentOperation. It was always failing, but we had commented it out because of race conditions, and so missed that this was an issue when it was added.

Could you elaborate please what exactly was failing and why?

---
In reply to: [526329457](https://github.com/dotnet/roslyn/pull/49453#discussion_r526329457) [](ancestors = 526329457)

333fred replied:
> Could you elaborate please what exactly was failing and why?

The constructor of `BlockOperation` calls `SetParentOperation`. Because of the race condition, we had commented out the verification code that ensured that the result of `SetParentOperation` was the node we were expecting. These statements already have parent operations, and so adding that verification back in failed here, because `SetParentOperation` did not set their parent nodes to this new Block.

---
In reply to: [526917375](https://github.com/dotnet/roslyn/pull/49453#discussion_r526917375) [](ancestors = 526917375,526329457)

AlekseyTs replied:
> The constructor of BlockOperation calls SetParentOperation

You mean for children that we are passing?

---
In reply to: [527277309](https://github.com/dotnet/roslyn/pull/49453#discussion_r527277309) [](ancestors = 527277309,526917375,526329457)

333fred replied:
> You mean for children that we are passing?

Yep. These children already have parents, so the Interlocked.CompareExchange does not set it to the new block and verification fails.

---
In reply to: [527279503](https://github.com/dotnet/roslyn/pull/49453#discussion_r527279503) [](ancestors = 527279503,527277309,526917375,526329457)

333fred replied:
I'm adding additional asserts to the CreateTemporaryBlock constructor to verify the scenario.

---
In reply to: [527283599](https://github.com/dotnet/roslyn/pull/49453#discussion_r527283599) [](ancestors = 527283599,527279503,527277309,526917375,526329457)

333fred commented on src/Compilers/Core/Portable/Operations/OperationMapBuilder.cs:
```diff
@@ -0,0 +1,62 @@
+        {
+            // In order to handle very large nested operators, we implement manual iteration here. Our operations are not order sensitive,
+            // so we don't need to maintain a stack, just iterate through every level. Visit above will have already recorded the given operation.
+            Debug.Assert(argument.ContainsKey(operation.Syntax) || operation.IsImplicit);
```
The test hook caught some overflows in the walker, so I included a standard manual recursion here.

333fred commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -352,6 +323,15 @@ Namespace Microsoft.CodeAnalysis.Operations
             End Select
         End Function
 
+        Public Function CreateFromArray(Of TBoundNode As BoundNode, TOperation As {Class, IOperation})(nodeArray As ImmutableArray(Of TBoundNode)) As ImmutableArray(Of TOperation)
```
Same deal: this is not modified, just moved.

333fred commented:
@AlekseyTs @dotnet/roslyn-compiler, this is ready for review.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationWalker.cs:
```diff
@@ -54,4 +54,57 @@ internal override void VisitNoneOperation(IOperation operation)
+            }
+        }
+
+        internal void VisitOperationArrayElement<T>(T operation, TArgument argument) where T : IOperation
```
>VisitOperationArrayElement [](start = 22, length = 26)

Is presence of this method necessary? It doesn't look like it is doing anything interesting or complicated.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationWalker.cs:
```diff
@@ -54,4 +54,57 @@ internal override void VisitNoneOperation(IOperation operation)
+    {
+        private int _recursionDepth;
+
+        internal void VisitArray<T>(IEnumerable<T> operations, TArgument argument) where T : IOperation
```
>VisitArray [](start = 22, length = 10)

Usage of "Array" in the name of this method is somewhat confusing. Consider renaming to "VisitOperations", "VisitSequence" or something similar

333fred replied:
I'm actually restructuring these in my `Children` refactoring PR (coming a bit later), and have already removed the `VisitOperationArrayElement` methods and renamed this method. I'll leave these until then.

---
In reply to: [526900387](https://github.com/dotnet/roslyn/pull/49453#discussion_r526900387) [](ancestors = 526900387)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationNodes.cs:
```diff
@@ -363,4 +363,20 @@ internal sealed partial class StaticLocalInitializationSemaphoreOperation
+        private BlockOperation(ImmutableArray<IOperation> statements, SemanticModel semanticModel, SyntaxNode syntax)
+            : base(semanticModel, syntax, isImplicit: true)
+        {
+            // Intentionally skipping SetParentOperation: this is used by CreateTemporaryBlock for the purposes of the
```
>Intentionally skipping SetParentOperation [](start = 15, length = 41)

What are the consequences of doing this? Certain things will not work/throw, etc. ?

333fred replied:
Added more detail to the comment.

---
In reply to: [526903983](https://github.com/dotnet/roslyn/pull/49453#discussion_r526903983) [](ancestors = 526903983)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationNodes.cs:
```diff
@@ -363,4 +363,20 @@ internal sealed partial class StaticLocalInitializationSemaphoreOperation
+
+    internal sealed partial class BlockOperation
+    {
+        public static BlockOperation CreateTemporaryBlock(ImmutableArray<IOperation> statements, SemanticModel semanticModel, SyntaxNode syntax)
```
>CreateTemporaryBlock [](start = 37, length = 20)

It would be good to explain what is special about this block that we call it "temporary". Are there restrictions on usage, etc. ?

333fred replied:
Added a doc comment.

---
In reply to: [526904956](https://github.com/dotnet/roslyn/pull/49453#discussion_r526904956) [](ancestors = 526904956)

AlekseyTs commented:
>         protected void SetParentOperation(IOperation? parent)

Consider asserting that parent is not s_unset here.

---
Refers to: src/Compilers/Core/Portable/Operations/Operation.cs:130 in 5a74cf2. [](commit_id = 5a74cf21ec028e4855410b6b47a1f7f44168b0bf, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/Operation.cs:
```diff
@@ -169,16 +158,14 @@ protected void SetParentOperation(IOperation? parent)
+                    VerifyParentOperation(parent, operation);
                     continue;
+#else
+                    break;
```
>break; [](start = 20, length = 6)

Why is it safe to skip remaining operations?

333fred replied:
Because this is entirely non-lazy now, there's no chance that we can observe an array where some of the elements were constructed before, and some later. They're all fresh nodes, so if one is set, they were all set. I actually think I can simplify the code more: nodes seen by `SetParentOperation` should never have existing parents, and this `if` is entirely unnecessary.

---
In reply to: [526911026](https://github.com/dotnet/roslyn/pull/49453#discussion_r526911026) [](ancestors = 526911026)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationMapBuilder.cs:
```diff
@@ -0,0 +1,62 @@
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class OperationMapBuilder : OperationWalker<Dictionary<SyntaxNode, IOperation>>
```
>OperationMapBuilder [](start = 26, length = 19)

Consider adding a doc comment explaining what kind of map is built by this type.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationMapBuilder.cs:
```diff
@@ -0,0 +1,62 @@
+                return null;
+            }
+
+            RecordOperation(operation, argument);
```
>RecordOperation(operation, argument); [](start = 12, length = 37)

I am not sure why we are making an assumption that every node goes through here. Nothing prevents me from calling specific visit method directly and bypassing this code.

AlekseyTs replied:
This can be somewhat mitigated by making the following changes:
- Do not implement walker on this type directly. Instead, have static method ```public void AddToMap(IOperation? operation, Dictionary<SyntaxNode, IOperation> argument)```
- The walker type should be private nested type with a singleton that is never exposed outside.
- AddToMap calls Visit on the singleton.

This way we are going to have a full control over how the visitor is used, but we will continue to be vulnerable to what base implementation is doing.

---
In reply to: [526924753](https://github.com/dotnet/roslyn/pull/49453#discussion_r526924753) [](ancestors = 526924753)

333fred replied:
I'll make those suggestions. Note that `OperationWalker` is a public type, so we need to keep the behavior consistent to avoid breaking our consumers as well.

---
In reply to: [526937934](https://github.com/dotnet/roslyn/pull/49453#discussion_r526937934) [](ancestors = 526937934,526924753)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationMapBuilder.cs:
```diff
@@ -0,0 +1,62 @@
+        {
+            // In order to handle very large nested operators, we implement manual iteration here. Our operations are not order sensitive,
+            // so we don't need to maintain a stack, just iterate through every level. Visit above will have already recorded the given operation.
+            Debug.Assert(argument.ContainsKey(operation.Syntax) || operation.IsImplicit);
```
>Debug.Assert(argument.ContainsKey(operation.Syntax) || operation.IsImplicit); [](start = 12, length = 77)

This doesn't look like a safe assumption because the method can be called directly.

333fred replied:
With the private walker change, this can no longer be called directly.

---
In reply to: [526926610](https://github.com/dotnet/roslyn/pull/49453#discussion_r526926610) [](ancestors = 526926610)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationMapBuilder.cs:
```diff
@@ -0,0 +1,62 @@
+                    Visit(operation.LeftOperand, argument);
+                    break;
+                }
+            } while (true);
```
>while (true); [](start = 14, length = 13)

Usage of ``` do {...} while (true)``` is somewhat unconventional. Consider using ```while (true) {...}``` of ```for(;;) {...}``` instead.

333fred replied:
For future reference, our formatter wants to make `for (;;)` into `for (; ; )`. I'll do the while version.

---
In reply to: [526930053](https://github.com/dotnet/roslyn/pull/49453#discussion_r526930053) [](ancestors = 526930053)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -29,10 +25,11 @@ internal abstract partial class MemberSemanticModel : CSharpSemanticModel
         private readonly CSharpSyntaxNode _root;
-        private readonly DiagnosticBag _ignoredDiagnostics = new DiagnosticBag();
-        private readonly ReaderWriterLockSlim _nodeMapLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+        private readonly DiagnosticBag _ignoredDiagnostics = new();
```
>new() [](start = 61, length = 5)

I do not believe this is a change for the better. One can no longer use a simple textual search for "new DiagnosticBag" to find all places allocating new bag, for example. IDE is not always there and when it is available, FindAllReferences can be extremely slow, especially when IDE just started.

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -29,10 +25,11 @@ internal abstract partial class MemberSemanticModel : CSharpSemanticModel
+        private readonly ReaderWriterLockSlim _nodeMapLock = new(LockRecursionPolicy.NoRecursion);
         // The bound nodes associated with a syntax node, from highest in the tree to lowest.
-        private readonly Dictionary<SyntaxNode, ImmutableArray<BoundNode>> _guardedNodeMap = new Dictionary<SyntaxNode, ImmutableArray<BoundNode>>();
+        private readonly Dictionary<SyntaxNode, ImmutableArray<BoundNode>> _guardedBoundNodeMap = new();
```
>new(); [](start = 98, length = 6)

Same comment here and below

AlekseyTs replied:
BTW, this changes like that do not follow a general guideline we have to not bundle unrelated refactorings and style changes with product changes in a single PR.

---
In reply to: [526947477](https://github.com/dotnet/roslyn/pull/49453#discussion_r526947477) [](ancestors = 526947477)

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -1191,23 +1204,27 @@ private CSharpSyntaxNode GetBindingRootOrInitializer(CSharpSyntaxNode node)
-            BoundNode result = highestBoundNode;
+            if (highestBoundNode is BoundGlobalStatementInitializer { Statement: var innerStatement })
+            {
+                // We don't have an IOperation node for the top-level statement here.
```
>top-level statement [](start = 60, length = 19)

When I read this, I think about regular top-level statements and immediately feel confused because I think that we cannot get here for them. It looks like this is very script specific and is not for a regular statement. Consider clarifying this.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1946,13 +1925,21 @@ private ILabeledOperation CreateBoundLabeledStatementOperation(BoundLabeledState
+            // If we're creating the tree for a speculatively-bound constructor initializer, there can be a bound sequence as the child node here
+            // that corresponds to the lifetime of any declared variables.
+            IOperation expression = Create(boundExpressionStatement.Expression);
+            if (boundExpressionStatement.Expression is BoundSequence sequence)
```
>boundExpressionStatement.Expression is BoundSequence sequence [](start = 16, length = 61)

Are we getting Operation.None in scenario like that. It feels like we shouldn't be getting any for the speculation scenario. Did we get the None node before the change? If, so, we should open an issue to follow up and probably change Implicit to false on that node rather than on this one.

333fred replied:
> Are we getting Operation.None in scenario like that.

Yes, we're producing a NoneOperation, and we got it before the change. The verification I added to `Scope_ConstructorInitializers_06` actually asserts on master, because it also has the bug this is asserting on. The main issue is that we have 3 nodes here:
1. The outer expression statement, here. This has the ConstructorInitializerSyntax as it's syntax
2. The middle NoneOperation. The syntax of this node in the operation tree is the ArgumentListSyntax under the ConstructorInitializerSyntax.
3. The bottom invocation. This is the node with conflicting syntax, it has the ConstructorInitializerSyntax as well. While this node feels like the right node to be explicit for this tree because it corresponds to the actual user-typed code, I can look into making this one explicit instead.

---
In reply to: [527022462](https://github.com/dotnet/roslyn/pull/49453#discussion_r527022462) [](ancestors = 527022462)

AlekseyTs replied:
> Yes, we're producing a NoneOperation, and we got it before the change. 

We discussed this offline and, given that, the problem with  ```isImplicit``` was present before and it is not about conflict with the NoneOperation created for the BoundSequence, the current implementation makes sense.

---
In reply to: [527230162](https://github.com/dotnet/roslyn/pull/49453#discussion_r527230162) [](ancestors = 527230162,527022462)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1946,13 +1925,21 @@ private ILabeledOperation CreateBoundLabeledStatementOperation(BoundLabeledState
+            if (boundExpressionStatement.Expression is BoundSequence sequence)
+            {
+                Debug.Assert(boundExpressionStatement.Syntax == sequence.Value.Syntax);
+                isImplicit = true;
```
>isImplicit = true; [](start = 16, length = 18)

Can we observe effect of this adjustments in any test that asserts the shape of the node we create?

333fred replied:
I can update the verification in `Scope_ConstructorInitializers_06` to go two nodes up, currently it's just using the regular machinery.

---
In reply to: [527029812](https://github.com/dotnet/roslyn/pull/49453#discussion_r527029812) [](ancestors = 527029812)

AlekseyTs replied:
> I can update the verification in Scope_ConstructorInitializers_06 to go two nodes up, currently it's just using the regular machinery.

As we discussed offline, it would be good to assert the shape of these node since they can be navigated to.

---
In reply to: [527230521](https://github.com/dotnet/roslyn/pull/49453#discussion_r527230521) [](ancestors = 527230521,527029812)

AlekseyTs commented:
Done with review pass (iteration 3)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [730497968](https://github.com/dotnet/roslyn/pull/49453#issuecomment-730497968) [](ancestors = 730497968)

AlekseyTs commented:
>         protected void SetParentOperation(IOperation? parent)

>

I was actually referring to the ```parent``` value.

---
In reply to: [730395296](https://github.com/dotnet/roslyn/pull/49453#issuecomment-730395296) [](ancestors = 730395296)

---
Refers to: src/Compilers/Core/Portable/Operations/Operation.cs:130 in 5a74cf2. [](commit_id = 5a74cf21ec028e4855410b6b47a1f7f44168b0bf, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/Operation.cs:
```diff
@@ -138,22 +129,22 @@ public string Language
 
         protected void SetParentOperation(IOperation? parent)
         {
+            Debug.Assert(Volatile.Read(ref _parentDoNotAccessDirectly) == s_unset);
```
>Debug.Assert(Volatile.Read(ref _parentDoNotAccessDirectly) == s_unset); [](start = 12, length = 71)

The fact that we are using ```Interlocked.CompareExchange``` below leads me to believe that a concurrency is possible. Given that, it feels like this assert can fail in some scenarios. If this method can no longer be called concurrently, we probably should adjust its implementation accordingly including the ```Volatile.Read``` usage.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/Operation.cs:
```diff
@@ -145,9 +136,7 @@ protected void SetParentOperation(IOperation? parent)
                 ((Operation)parent).OwningSemanticModel == null || OwningSemanticModel == null);
```
>// explicit cast is not allowed, so using "as" instead [](start = 12, length = 54)

Is this comment no longer relevant?

333fred replied:
(Note to future self, codeflow is showing this wrong, it should be on the comment removal in SetParentOperation below)

That comment felt pretty self-explanatory to me, explaining the what without the why.

---
In reply to: [527721421](https://github.com/dotnet/roslyn/pull/49453#discussion_r527721421) [](ancestors = 527721421)

AlekseyTs commented:
Done with review pass (iteration 5).

333fred commented:
@AlekseyTs addressed feecback.

---
In reply to: [731214601](https://github.com/dotnet/roslyn/pull/49453#issuecomment-731214601) [](ancestors = 731214601)

AlekseyTs reviewed (approved):
LGTM (iteration 6)

jcouv commented on src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:
```diff
@@ -32,7 +28,8 @@ internal abstract partial class MemberSemanticModel : CSharpSemanticModel
         // The bound nodes associated with a syntax node, from highest in the tree to lowest.
-        private readonly Dictionary<SyntaxNode, ImmutableArray<BoundNode>> _guardedNodeMap = new Dictionary<SyntaxNode, ImmutableArray<BoundNode>>();
+        private readonly Dictionary<SyntaxNode, ImmutableArray<BoundNode>> _guardedBoundNodeMap = new Dictionary<SyntaxNode, ImmutableArray<BoundNode>>();
+        private readonly Dictionary<SyntaxNode, IOperation> _guardedIOperationNodeMap = new Dictionary<SyntaxNode, IOperation>();
```
>new Dictionary<SyntaxNode, IOperation>(); [](start = 88, length = 41)

nit: consider using singleton as sentinel value

jcouv commented on src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -1,11 +1,15 @@
 Microsoft.CodeAnalysis.GeneratorAttribute.GeneratorAttribute(string firstLanguage, params string[] additionalLanguages) -> void
 Microsoft.CodeAnalysis.GeneratorAttribute.Languages.get -> string[]
 Microsoft.CodeAnalysis.ITypeSymbol.IsRecord.get -> bool
+Microsoft.CodeAnalysis.Operations.OperationWalker<TArgument>
```
Is making this new API public related to removing laziness from implementation? Feels like this should be kept separate.
If we add a public API, we'd want direct validation on it.

jcouv reviewed (approved):
LGTM Thanks (iteration 6)
